optuna/study/_multi_objective.py

```diff
@@ -69,6 +74,64 @@ def _get_pareto_front_trials(study: "optuna.study.Study") -> List[FrozenTrial]:
     return _get_pareto_front_trials_by_trials(study.trials, study.directions)
 
 
+def _fast_non_dominated_sort(
+    objective_values: np.ndarray,
+    *,
+    penalty: np.ndarray | None = None,
+    n_below: int | None = None,
+) -> np.ndarray:
```

Comment:
```python
from __future__ import annotations

from collections import defaultdict
import time

import numpy as np


def run_alnusjaponica(objective_values: np.ndarray) -> np.ndarray:
    domination_mat = np.all(
        objective_values[:, np.newaxis, :] >= objective_values[np.newaxis, :, :], axis=2
    ) & np.any(
        objective_values[:, np.newaxis, :] > objective_values[np.newaxis, :, :], axis=2
    )

    domination_list = np.nonzero(domination_mat)
    domination_map = defaultdict(list)
    for dominated_idx, dominating_idx in zip(*domination_list):
        domination_map[dominating_idx].append(dominated_idx)

    ranks = np.full(len(objective_values), -1)
    dominated_count = np.sum(domination_mat, axis=1)

    rank = -1
    ranked_idx_num = 0
    while ranked_idx_num < len(objective_values):
        (non_dominated_idxs,) = np.nonzero(dominated_count == 0)
        ranked_idx_num += len(non_dominated_idxs)
        rank += 1
        ranks[non_dominated_idxs] = rank

        dominated_count[non_dominated_idxs] = -1
        for non_dominated_idx in non_dominated_idxs:
            dominated_count[domination_map[non_dominated_idx]] -= 1
    return ranks


def is_pareto_front_nd(ordered_objective_values: np.ndarray) -> np.ndarray:
    loss_values = ordered_objective_values.copy()
    n_trials = loss_values.shape[0]
    on_front_indices = np.arange(n_trials)
    next_index = 0
    while next_index < len(loss_values):
        nondominated_mask = np.any(loss_values < loss_values[next_index], axis=1)
        nondominated_mask[~nondominated_mask] |= np.all(
            loss_values[~nondominated_mask] == loss_values[next_index], axis=1
        )
        nondominated_mask[next_index] = True
        # Remove dominated trials.
        on_front_indices = on_front_indices[nondominated_mask]
        loss_values = loss_values[nondominated_mask]
        next_index = np.sum(nondominated_mask[:next_index]) + 1

    mask = np.zeros(n_trials, dtype=bool)
    mask[on_front_indices] = True
    return mask


def is_pareto_front_2d(ordered_objective_values: np.ndarray) -> np.ndarray:
    n_trials = ordered_objective_values.shape[0]
    cummin_value1 = np.minimum.accumulate(ordered_objective_values[:, 1])
    is_value1_min = cummin_value1 == ordered_objective_values[:, 1]
    is_value1_new_min = cummin_value1[1:] < cummin_value1[:-1]

    on_front = np.ones(n_trials, dtype=bool)
    is_value0_same = ordered_objective_values[1:, 0] == ordered_objective_values[:-1, 0]
    on_front[1:] = is_value1_min[1:] & (is_value0_same | is_value1_new_min)
    return on_front


def is_pareto_front(ordered_objective_values: np.ndarray) -> np.ndarray:
    (n_trials, n_objectives) = ordered_objective_values.shape
    if n_objectives == 1:
        return ordered_objective_values[:, 0] == ordered_objective_values[0]
    elif n_objectives == 2:
        return is_pareto_front_2d(ordered_objective_values)
    else:
        return is_pareto_front_nd(ordered_objective_values)


def run_nabenabe(objective_values: np.ndarray) -> np.ndarray:
    (n_trials, n_objectives) = objective_values.shape
    ranks = np.zeros(n_trials, dtype=int)
    rank = 0
    indices = np.arange(n_trials)

    if n_objectives == 1:
        order = np.argsort(objective_values[:, 0])
    else:
        order = np.lexsort([objective_values[:, 1], objective_values[:, 0]])

    order_inv = np.zeros_like(order)
    order_inv[order] = np.arange(n_trials)
    ordered_objective_values = objective_values[order]
    while indices.size > 0:
        on_front = is_pareto_front(ordered_objective_values)
        ranks[indices[on_front]] = rank
        # Remove pareto front points.
        indices = indices[~on_front]
        ordered_objective_values = ordered_objective_values[~on_front]
        rank += 1

    return ranks[order_inv]


def measure_time(target, objective_values: np.ndarray) -> tuple[np.ndarray, float]:
    start = time.time()
    results = target(objective_values.copy())
    elapsed_time = (time.time() - start) * 1000
    # print(f"{elapsed_time:.2f}[ms]")
    return results, elapsed_time


if __name__ == "__main__":
    n_trials = 2000
    n_objectives = 2
    n_seeds = 5
    results = {"nabenabe": [], "alnusjaponica": []}
    for seed in range(n_seeds):
        rng = np.random.RandomState(seed)
        objective_values = rng.normal(size=(n_trials, n_objectives))
        ans_nabenabe, t = measure_time(run_nabenabe, objective_values)
        results["nabenabe"].append(t)
        ans_alnusjaponica, t = measure_time(run_alnusjaponica, objective_values)
        results["alnusjaponica"].append(t)
        print(np.all(ans_alnusjaponica == ans_nabenabe))

    print({k: f"{np.mean(v):.2f} +/- {np.std(v) / np.sqrt(n_seeds):.2f} [ms]" for k, v in results.items()})

```

For simplicity, I removed `penalty` and `n_below`, but this implementation maintains the same results by your program yet much quicker.

Reply:
The unit is milliseconds.
||Mine|Yours|
|:--:|:--:|:--:|
|`n_trials=1000, n_objectives=2`|1.04 $\pm$ 0.1|84.5 $\pm$ 3.1|
|`n_trials=10000, n_objectives=2`|16.4 $\pm$ 1.4|9858.0 $\pm$ 126.8|
|`n_trials=1000, n_objectives=3`|17.54 $\pm$ 2.1|61.5 $\pm$ 1.0|
|`n_trials=10000, n_objectives=3`|469.0 $\pm$ 10.1|7025.0 $\pm$ 181.4|

Reply:
I think your implementation of dominance with `penalty` includes the unnecessary consideration of `penalty` for feasible cases.
Namely, we do not consider the penalty amount once each trial satisfies the penalty, but yours are considering the penalty amount even for feasible cases.

Please check the following definition by Deb et al. [1]:

![image](https://github.com/optuna/optuna/assets/47781922/9c904135-d5a9-415a-997d-786ecbe318ce)


Plus, the implementation below is much quicker.

```python
def calculate_nondomination_rank_with_penalty(objective_values: np.ndarray, penalty: np.ndarray | None = None) -> np.ndarray:
    if penalty is None:
        return calculate_nondomination_rank(objective_values)

    # If values[i] constrained-dominates values[j] given penalty[i] and penalty[j],
    # one of the following must be satisfied:
    # 1. penalty[i] <= 0 and penalty[j] > 0,
    # 2. penalty[i] > 0 and penalty[j] > 0 and penalty[i] < penalty[j], or
    # 3. penalty[i] <= 0 and penalty[j] <= 0 and values[i] dominates values[j].
    # Therefore, if trials[i] is feasible and trials[j] is infeasible,
    # nondomination_rank[i] <= nondomination_rank[j] always holds and we can separate the sortings
    # for feasible trials and infeasible trials.
    # Ref: Definition 1 by K. Deb et al. in `A Fast and Elitist Multiobjective Genetic Algorithm: NSGA-II`
    if len(penalty) != len(objective_values):
        raise ValueError(
            f"The length of penalty and objective_values must be same, but got "
            f"len(penalty)={len(penalty)} and len(objective_values)={len(objective_values)}."
        )

    penalty[np.isnan(penalty)] = np.inf
    is_feasible = penalty <= 0
    nondomination_rank = np.zeros(len(objective_values), dtype=int)
    nondomination_rank[is_feasible] += calculate_nondomination_rank(objective_values[is_feasible])
    nondomination_rank[~is_feasible] += np.max(nondomination_rank[is_feasible], initial=-1) + 1
    nondomination_rank[~is_feasible] += calculate_nondomination_rank(
        penalty[~is_feasible, np.newaxis],
    )
    return nondomination_rank
```

[1] K. Deb et al. A Fast and Elitist Multiobjective Genetic Algorithm: NSGA-II

Reply:
> this implementation maintains the same results as your program yet much quicker.

Thanks for the suggestion. I just move what's implemented in `_tpe/sampler.py` and no need to stick to the first implementation. I'll consider how to reconcile current `_get_pareto_front_trials_by_trials` function with your suggestion.
> I think your implementation of dominance with penalty includes the unnecessary consideration of penalty for feasible cases.

In my understanding, the penalty is set to 0 when the trial is feasible, thus having no influence on the result. Anyway, I'll use the faster implementation in the suggestion.

Reply:
```python
from __future__ import annotations

import itertools

import numpy as np


def is_pareto_front(trials: list[Trial]) -> np.ndarray:
    n_trials = len(trials)
    is_front = np.zeros(n_trials, dtype=bool)
    next_index = 0
    nondominated_indices = np.arange(n_trials)
    while next_index < len(trials):
        nondominated_mask = np.array([not dominates(t, trials[next_index]) for t in trials])
        trials = list(itertools.compress(trials, nondominated_mask))
        nondominated_indices = nondominated_indices[nondominated_mask]
        next_index = np.sum(nondominated_mask[:next_index]) + 1

    is_front[nondominated_indices] = True
    return is_front


def calculate_nondomination_rank(trials: list[Trial]) -> np.ndarray:
    n_trials = len(trials)
    ranks = np.zeros(n_trials, dtype=int)
    rank = 0
    indices = np.arange(n_trials)
    while indices.size > 0:
        on_front = is_pareto_front(trials)
        ranks[indices[on_front]] = rank
        indices = indices[~on_front]
        trials = list(itertools.compress(trials, ~on_front))
        rank += 1

    return ranks
```

This is Just a memo for future discussion, please ignore it for now.

When using `dominates`, the runtime will be much much longer compared to the vectorization version, but still it runs quicker than creating the dominance matrix.